### ISSUE TITLE
GenAPI issues generate invalid C# for internal types

### ISSUE THREAD

**Issue by JunTaoLuo** [dotnet/arcade#4488]
We have encountered cases where we want to generate internal members in reference assemblies. It would be nice to be able to distinguish between private and internal members instead of the only `--all` option that accounts for both. However, given the current implementation, here are a few issues we have ran into. FYI @dougbu @wtgodbe @safern @tannergooding 

1. Null Argument Exceptions

For example, when generating for the field `F:Microsoft.AspNetCore.Mvc.ApplicationModels.ActionAttributeRouteModel.<GetAttributeRoutes>d__3.{}2__current`
the following exception will be hit
```
Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'source')
     at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
     at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
     at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
     at Microsoft.Cci.Extensions.CSharp.CSharpCciExtensions.GetAttributeOfType(IEnumerable`1 attributes, String attributeName) in /_/src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs:line 666
     at Microsoft.Cci.Extensions.CSharp.CSharpCciExtensions.GetValueTupleNames(IEnumerable`1 attributes) in /_/src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs:line 683
     at Microsoft.Cci.Writers.CSharp.CSDeclarationWriter.WriteTypeName(ITypeReference type, Boolean noSpace, IEnumerable`1 attributes, Boolean useTypeKeywords, Boolean omitGenericTypeList) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.cs:line 272
     at Microsoft.Cci.Writers.CSharp.CSDeclarationWriter.WriteFieldDefinition(IFieldDefinition field) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Fields.cs:line 56
     at Microsoft.Cci.Writers.CSharp.CSDeclarationWriter.WriteMemberDeclaration(ITypeDefinitionMember member) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.cs:line 180
     at Microsoft.Cci.Writers.CSharp.CSDeclarationWriter.WriteDeclaration(IDefinition definition) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.cs:line 120
     at Microsoft.Cci.Writers.CSharpWriter.Visit(ITypeDefinitionMember member) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 244
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(IEnumerable`1 members) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 104
     at Microsoft.Cci.Writers.CSharpWriter.Visit(IEnumerable`1 members) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 147
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(ITypeDefinition parentType, IEnumerable`1 fields) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 117
     at Microsoft.Cci.Writers.CSharpWriter.Visit(ITypeDefinition parentType, IEnumerable`1 fields) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 211
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(ITypeDefinition type) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 74
     at Microsoft.Cci.Writers.CSharpWriter.Visit(ITypeDefinition type) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 137
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(IEnumerable`1 types) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 62
     at Microsoft.Cci.Writers.CSharpWriter.Visit(IEnumerable`1 types) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 116
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(ITypeDefinition type) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 79
     at Microsoft.Cci.Writers.CSharpWriter.Visit(ITypeDefinition type) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 137
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(IEnumerable`1 types) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 62
     at Microsoft.Cci.Writers.CSharpWriter.Visit(IEnumerable`1 types) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 116
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(INamespaceDefinition ns) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 55
     at Microsoft.Cci.Writers.CSharpWriter.Visit(INamespaceDefinition ns) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 105
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(IEnumerable`1 namespaces) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 43
     at Microsoft.Cci.Traversers.SimpleTypeMemberTraverser.Visit(IAssembly assembly) in /_/src/Microsoft.Cci.Extensions/Traversers/SimpleTypeMemberTraverser.cs:line 35
     at Microsoft.Cci.Writers.CSharpWriter.Visit(IAssembly assembly) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 91
     at Microsoft.Cci.Writers.CSharpWriter.WriteAssemblies(IEnumerable`1 assemblies) in /_/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs:line 79
     at Microsoft.DotNet.GenAPI.Program.<>c__DisplayClass0_0.<Main>b__0() in /_/src/Microsoft.DotNet.GenAPI/Program.cs:line 97
     at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
     at Microsoft.DotNet.GenAPI.Program.Main(String[] args) in /_/src/Microsoft.DotNet.GenAPI/Program.cs:line 165
```

2. Anonymous classes

Anonymous classes are generated when the --all option is specified and the code is not compilable. We have to manually remove the anonymous classes:
```c#
    public static partial class DataProtectionProvider
    {
        public static Microsoft.AspNetCore.DataProtection.IDataProtectionProvider Create(System.IO.DirectoryInfo keyDirectory) { throw null; }
        public static Microsoft.AspNetCore.DataProtection.IDataProtectionProvider Create(System.IO.DirectoryInfo keyDirectory, System.Action<Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder> setupAction) { throw null; }
        public static Microsoft.AspNetCore.DataProtection.IDataProtectionProvider Create(System.IO.DirectoryInfo keyDirectory, System.Action<Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder> setupAction, System.Security.Cryptography.X509Certificates.X509Certificate2 certificate) { throw null; }
        public static Microsoft.AspNetCore.DataProtection.IDataProtectionProvider Create(System.IO.DirectoryInfo keyDirectory, System.Security.Cryptography.X509Certificates.X509Certificate2 certificate) { throw null; }
        public static Microsoft.AspNetCore.DataProtection.IDataProtectionProvider Create(string applicationName) { throw null; }
        public static Microsoft.AspNetCore.DataProtection.IDataProtectionProvider Create(string applicationName, System.Security.Cryptography.X509Certificates.X509Certificate2 certificate) { throw null; }
        internal static Microsoft.AspNetCore.DataProtection.IDataProtectionProvider CreateProvider(System.IO.DirectoryInfo keyDirectory, System.Action<Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder> setupAction, System.Security.Cryptography.X509Certificates.X509Certificate2 certificate) { throw null; }
        [System.Runtime.CompilerServices.CompilerGeneratedAttribute]
        private sealed partial class <>c
        {
            public static readonly Microsoft.AspNetCore.DataProtection.DataProtectionProvider.<>c __9;
            public static System.Action<Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder> __9__1_0;
            public static System.Action<Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder> __9__4_0;
            public <>c() { }
            private static void .cctor() { }
            internal void <Create>b__1_0(Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder builder) { }
            internal void <Create>b__4_0(Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder builder) { }
        }
        [System.Runtime.CompilerServices.CompilerGeneratedAttribute]
        private sealed partial class <>c__DisplayClass0_0
        {
            public string applicationName;
            public <>c__DisplayClass0_0() { }
            internal void <Create>b__0(Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder builder) { }
        }
        [System.Runtime.CompilerServices.CompilerGeneratedAttribute]
        private sealed partial class <>c__DisplayClass3_0
        {
            public string applicationName;
            public <>c__DisplayClass3_0() { }
            internal void <Create>b__0(Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder builder) { }
        }
    }
```

3. Private static constructors

Non-compilable private static constructor are generated, which we have to remove manually:

```C#
    internal static partial class EncodingUtil
    {
        public static readonly System.Text.UTF8Encoding SecureUtf8Encoding;
        private static void .cctor() { }
    }
```

4. Unsafe interfaces missing the unsafe keyword:

Implementation C# source:
```C#
    internal unsafe interface IBCryptGenRandom
    {
        void GenRandom(byte* pbBuffer, uint cbBuffer);
    }
```

Generated reference C# source
```C#
    internal partial interface IBCryptGenRandom
    {
        void GenRandom(byte* pbBuffer, uint cbBuffer);
    }
```

5. Nullable types not generated

It seems like some nullable types are not generated as such by GenAPI. It doesn't happen for all nullable types so it's unclear what exactly causes the errors.

Implementation C# source:
```C#

namespace Microsoft.AspNetCore.Mvc.Infrastructure
{
#nullable enable
    internal abstract class ResourceInvoker
    {
        protected readonly DiagnosticListener _diagnosticListener;
        protected readonly ILogger _logger;
        protected readonly IActionContextAccessor _actionContextAccessor;
        protected readonly IActionResultTypeMapper _mapper;
        protected readonly ActionContext _actionContext;
        protected readonly IFilterMetadata[] _filters;
        protected readonly IList<IValueProviderFactory> _valueProviderFactories;

        private AuthorizationFilterContextSealed? _authorizationContext;
        private ResourceExecutingContextSealed? _resourceExecutingContext;
        private ResourceExecutedContextSealed? _resourceExecutedContext;
        private ExceptionContextSealed? _exceptionContext;
        private ResultExecutingContextSealed? _resultExecutingContext;
        private ResultExecutedContextSealed? _resultExecutedContext;

        // Do not make this readonly, it's mutable. We don't want to make a copy.
        // https://blogs.msdn.microsoft.com/ericlippert/2008/05/14/mutating-readonly-structs/
        protected FilterCursor _cursor;
        protected IActionResult? _result;
        protected object? _instance;
...
}
```

Generated reference C# source
```C#

    [System.Runtime.CompilerServices.NullableAttribute((byte)0)]
    [System.Runtime.CompilerServices.NullableContextAttribute((byte)1)]
    internal abstract partial class ResourceInvoker
    {
        protected readonly Microsoft.AspNetCore.Mvc.ActionContext _actionContext;
        protected readonly Microsoft.AspNetCore.Mvc.Infrastructure.IActionContextAccessor _actionContextAccessor;
        [System.Runtime.CompilerServices.NullableAttribute((byte)2)]
        private Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.AuthorizationFilterContextSealed _authorizationContext;
        protected Microsoft.AspNetCore.Mvc.Filters.FilterCursor _cursor;
        protected readonly System.Diagnostics.DiagnosticListener _diagnosticListener;
        [System.Runtime.CompilerServices.NullableAttribute((byte)2)]
        private Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ExceptionContextSealed _exceptionContext;
        protected readonly Microsoft.AspNetCore.Mvc.Filters.IFilterMetadata[] _filters;
        [System.Runtime.CompilerServices.NullableAttribute((byte)2)]
        protected object _instance;
        protected readonly Microsoft.Extensions.Logging.ILogger _logger;
        protected readonly Microsoft.AspNetCore.Mvc.Infrastructure.IActionResultTypeMapper _mapper;
        [System.Runtime.CompilerServices.NullableAttribute((byte)2)]
        private Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResourceExecutedContextSealed _resourceExecutedContext;
        [System.Runtime.CompilerServices.NullableAttribute((byte)2)]
        private Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResourceExecutingContextSealed _resourceExecutingContext;
        [System.Runtime.CompilerServices.NullableAttribute((byte)2)]
        protected Microsoft.AspNetCore.Mvc.IActionResult _result;
        [System.Runtime.CompilerServices.NullableAttribute((byte)2)]
        private Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultExecutedContextSealed _resultExecutedContext;
        [System.Runtime.CompilerServices.NullableAttribute((byte)2)]
        private Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultExecutingContextSealed _resultExecutingContext;
        protected readonly System.Collections.Generic.IList<Microsoft.AspNetCore.Mvc.ModelBinding.IValueProviderFactory> _valueProviderFactories;
...
}
```
In this case, it looks like NullableAttribute is added to the nullable types but the types are not nullable themselves. Is this an oversight or a feature that we are misunderstanding?

To repro these issues, find the implementation projects containing the code I've pasted here and run `dotnet msbuild /t:GenerateReferenceSource` to run genAPI to produce the generated code.

**Comment by safern** [comment]
cc: @ericstj

**Comment by ericstj** [comment]
GenAPI has never before been hardened to the task of emittting C# for all APIs so I'm not surprised that these issues exist.  It's not something we've needed it to do.

For the first two issues I think it makes sense to filter out compiler-generated types.  Those shouldn't be emitted if the output is C#.  @dougbu was chatting before about wanting more flexibility in filtering.  I think the work here would be to work on more filters [here](https://github.com/dotnet/arcade/tree/master/src/Microsoft.Cci.Extensions/Filters) and plumb them through GenAPI commandline.

3 looks just like a bug here: https://github.com/dotnet/arcade/blob/63cb4d2c2b1adae0ae74e552beccdbd63af400a7/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Methods.cs#L97-L123

4 is likely a feature cab that we haven't noticed yet and would require work here: https://github.com/dotnet/arcade/blob/63cb4d2c2b1adae0ae74e552beccdbd63af400a7/src/Microsoft.Cci.Extensions/Writers/CSharp/CSDeclarationWriter.Types.cs#L192-L196

@JunTaoLuo is this something you'd consider helping address to avoid some of your workarounds?

**Comment by dougbu** [comment]
Thanks for the pointers @ericstj❕ We've also chatted about GenAPI being open source 😃

@JunTaoLuo note some of this would get nicely papered over if `private` members were excluded using a new `--internal` command-line option and related filter instead of `--all`.

**Comment by dougbu** [comment]
Separately, @jaredpar does Roslyn evaulate the `sizeof()` operator at compile time when applied to a user `struct`? If yes, GenAPI's handling of `struct` fields is almost always incorrect and all `struct` fields should _always_ be generated (regardless of their visibility modifiers).

If not, there's still a GenAPI problem in this area because it generates `__dummy` and `__dummyPrimitive` _instead of_ `public` `struct` fields. E.g. (with `--all` in this case) for
``` c#
namespace Microsoft.AspNetCore.Cryptography
{
    // http://msdn.microsoft.com/en-us/library/windows/desktop/aa381414(v=vs.85).aspx
    [StructLayout(LayoutKind.Sequential)]
    internal unsafe struct DATA_BLOB
    {
        public uint cbData;
        public byte* pbData;
    }
}
```
GenAPI generates
``` c#
namespace Microsoft.AspNetCore.Cryptography
{
    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
    internal partial struct DATA_BLOB
    {
        private int _dummyPrimitive;
    }
}
```

**Comment by ericstj** [comment]
`sizeof()` becomes IL `sizeof` instruction, executed at runtime.  We've discussed the use of the `_dummy` fields at length (and Jared has a [good blog on topic](https://blog.paranoidcoding.com/2016/02/15/are-private-members-api-surface.html)) so we should avoid repeating that discussion unless there is some observed problem with the approach.

The stripping of fields when using `--all` does seem to be a bug.  Let's call that issue 5.  Problem is here: https://github.com/dotnet/arcade/blob/14abaee3dba41fbe608431fb6a4f4b2435dcac33/src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs#L158-L223  It assumes a public `IsVisibleOutsideAssembly` in addition to filter.  Not sure why its doing that.  It's done so [since this was added](https://github.com/dotnet/buildtools/commit/f2b02029f45cea37ed4e0b4889e5c8b02d9f75e7), but it should probably just use the filter to decide what fields to write.

**Comment by dougbu** [comment]
Had an offline conversation w/ @jaredpar. The `sizeof(answer)` is "no, that operator is evaluated at runtime except in the few documented cases e.g. `sizeof(int)`. So, we only have a problem w/ the missing `public` (and `internal`) fields.

**Comment by dougbu** [comment]
@ericstj sorry my [last comment](https://github.com/dotnet/arcade/issues/4488#issuecomment-562299347) crossed w/ [yours](https://github.com/dotnet/arcade/issues/4488#issuecomment-562276315). I agree generating `__dummy` et cetera is warranted and mentioned them only to introduce my "_instead of_" line in my [comment before that one](https://github.com/dotnet/arcade/issues/4488#issuecomment-562253840).

And, thanks for the additional pointer.

**Comment by JunTaoLuo** [comment]
@ericstj I added another question (5) in the initial comment. Would like to clarify what the expected behaviour is for nullable types.

**Comment by dougbu** [comment]
@JunTaoLuo with the missing `public` fields in `struct`s, we're up to issue (6) 😺

**Comment by dougbu** [comment]
All, @jaredpar and I had an offline conversation about some of this the other day. He suggested GenAPI should notice when the image it's scanning contains _any_ `[InternalsVisibleTo]` attributes. In that case, it should include all `internal` types and members. Might want an `--ignore-internals` command line option for legacy behaviour i.e. continue to support incorrect (or at least incomplete) ref/ assemblies.

If we had that, ASP.NET Core would be able to remove most of our *.Manual.cs files and narrow them down to just the few types we need to exclude from regular generation.

This sounds like a great idea to me. If it's too much work, we could use an `--include-internals` option for the short term.

----
Note: I'm not suggesting trying to switch gears (again) for 3.0.2 work in ASP.NET Core repos.

**Comment by ericstj** [comment]
For Nullable this was all a recent addition made by @safern.  You are probably using an old version of GenAPI.

@dougbu it's reasonable to have a new filter that is public + internals-only-when-IVT is used.  This shouldn't be the default public filter though since that would break folks who use GenAPI and only use IVT for runtime assemblies and testing.

**Comment by jaredpar** [comment]
@ericstj 

> This shouldn't be the default public filter though since that would break folks who use GenAPI and only use IVT for runtime assemblies and testing.

While that's true including internals when `[InternalsVisibleTo]` is the correct behavior. To exclude `internal` members in that scenario is to create a reference assembly that is not representative of the implementation assembly that it is run against. 

I understand that compatibility is important here but how will you advertise to customers that the correct behavior can only be achieved by opting into a switch? Conversely how many customers are actually depending on the incorrect assemblies being produced? Maybe it's simpler to just update them.

**Comment by ericstj** [comment]
The purpose of public only filter is **public only**.  I'm ok with another filter that matches what folks desire the default to be, but it shouldn't be the public only filter, and we plan to continue to use the public only filter for .NET libraries as we have no intent of growing our public facing reference assemblies with internal surface area.  I'm not telling anyone no here I'm just describing how this work should proceed if folks wish to pick it up without disrupting consumers of this tool.

**Comment by dougbu** [comment]
I'm self-assigning for at least part of this work. Will be experimenting and coding&hellip;

**Comment by dougbu** [comment]
@jaredpar @ericstj I could use some help w/ problem (4).

GenAPI correctly marks delegates and methods as `unsafe` when appropriate. But, it never does so for a `class`, `interface`, or `struct`. Adding
``` c#
if (type.IsUnsafeType())
{
    WriteKeyword("unsafe");
}
```
does nothing because the `TypeCode` of `unsafe` types is _not_ `PrimitiveTypeCode.Pointer`. Instead it's `NotPrimitive` in the cases I checked i.e. perfectly normal.

How are non-primitive `unsafe` types identified in CCI? If that's a missing CCI feature, what distinguishes these types in the IL?

**Comment by dougbu** [comment]
Nevermind, took a step away to walk the dog and figured it out: Real problem is that `WriteMethodDefinition(...)` in CSDeclarationWriter.Methods.cs limited the modifier keywords applied to methods in `interface`s too much. While `IBCryptGenRandom` is marked `unsafe` in @JunTaoLuo's example above, it's sufficient to apply the keyword to its `GenRandom(...)` method.

I made a fix and the generated code is now
``` c#
internal partial interface IBCryptGenRandom
{
    unsafe void GenRandom(byte* pbBuffer, uint cbBuffer);
}
```

**Comment by dougbu** [comment]
All but "Nullable types not generated" addressed in d6dd7a000fe9. Already have a number of issues related to nullability elsewhere in this repo and that point feels redundant.
